components/brave_wallet/browser/solana_instruction.cc

```diff
@@ -69,95 +51,118 @@ bool SolanaInstruction::operator==(const SolanaInstruction& ins) const {
+//
 // static
-absl::optional<SolanaInstruction> SolanaInstruction::Deserialize(
-    const std::vector<SolanaAccountMeta>& message_account_metas,
-    const std::vector<uint8_t>& bytes,
-    size_t* bytes_index) {
-  DCHECK(bytes_index);
-  if (*bytes_index >= bytes.size())
-    return absl::nullopt;
-
-  uint8_t program_id_index = bytes[(*bytes_index)++];
-  if (program_id_index >= message_account_metas.size())
+absl::optional<SolanaInstruction> SolanaInstruction::FromCompiledInstruction(
+    const SolanaCompiledInstruction& compiled_instruction,
+    const SolanaMessageHeader& message_header,
+    const std::vector<SolanaAddress>& static_accounts,
+    const std::vector<SolanaMessageAddressTableLookup>& addr_table_lookups,
+    uint8_t num_of_write_indexes,
+    uint8_t num_of_read_indexes) {
+  int num_writable_signed_accounts =
+      message_header.num_required_signatures -
+      message_header.num_readonly_signed_accounts;
+  int num_writable_unsigned_accounts =
+      static_accounts.size() - message_header.num_required_signatures -
+      message_header.num_readonly_signed_accounts;
+  if (num_writable_signed_accounts < 0 ||
+      num_writable_unsigned_accounts < 0) {  // invalid message header
     return absl::nullopt;
-  const std::string program_id = message_account_metas[program_id_index].pubkey;
+  }
 
-  auto ret = CompactU16Decode(bytes, *bytes_index);
-  if (!ret)
+  // Program ID of compiled_instruction should be in static accounts.
+  if (compiled_instruction.program_id_index() >= static_accounts.size()) {
     return absl::nullopt;
-  const uint16_t num_of_accounts = std::get<0>(*ret);
-  *bytes_index += std::get<1>(*ret);
+  }
+  std::string program_id =
+      static_accounts[compiled_instruction.program_id_index()].ToBase58();
 
   std::vector<SolanaAccountMeta> account_metas;
-  for (size_t i = 0; i < num_of_accounts; ++i) {
-    if (*bytes_index >= bytes.size())
-      return absl::nullopt;
-    uint8_t account_index = bytes[(*bytes_index)++];
-    if (account_index >= message_account_metas.size())
+  for (const auto& account_index : compiled_instruction.account_indexes()) {
+    std::string account_key;
+    absl::optional<uint8_t> address_table_lookup_index = absl::nullopt;
+    bool is_signer = false;
+    bool is_writable = false;
+
+    if (account_index < static_accounts.size()) {  // static accounts
+      account_key = static_accounts[account_index].ToBase58();
+      is_signer = account_index < message_header.num_required_signatures;
+      is_writable =
+          is_signer ? account_index < num_writable_signed_accounts
+                    : account_index - message_header.num_required_signatures <
+                          num_writable_unsigned_accounts;
+    } else if (account_index <
+               (static_accounts.size() + num_of_write_indexes +
+                num_of_read_indexes)) {  // dynamic loaded accounts
+      // Keep is_signer as false as all signers should be in static accounts.
```

Comment:
can we add this [link](https://docs.solana.com/proposals/versioned-transactions#limitations) to this comment?
so we know the reason why signers have to be in static account.
```
Transaction signers may not be loaded through an address lookup table, the full address of each signer must be serialized in the transaction. This ensures that the performance of transaction signature checks is not affected.
```

Reply:
46f9f027bd45d2f03324c63a9724ddbbe1812de0